File: app/utilities.py
import logging
import sys

from pathlib import Path

from app_config import AppConfig


def link_source_to_destination(source_path: Path, destination_path: Path, logger: logging.Logger) -> None:
    """ Hardlinks the source file to the destination. """
    destination_parent = destination_path.parent
    if not destination_parent.exists():
        destination_parent.mkdir(parents=True, exist_ok=True)
        logger.info(f'Created parent directory {destination_parent} for {destination_path}.')

    if destination_path.exists():
        # we don't want to overwrite existing files
        return

    try:
        destination_path.hardlink_to(source_path)
        logger.info(f'Linked {source_path} to {destination_path}')
    except FileExistsError:
        return
    except OSError as e:
        logger.error(f'Error linking {source_path} to {destination_path}: {e}')
        return

def prepare_logger() -> logging.Logger:
    """ Prepares the logger for the application. """
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(levelname)s - %(message)s',
        stream=sys.stderr  # stderr is unbuffered!
    )
    return logging.getLogger(__name__)

def source_path_is_qualified(source_path: Path, app_config: AppConfig, logger: logging.Logger) -> bool:
    """ Checks whether the given path is qualified for linking. """
    if not source_path or not source_path.exists():
        logger.info(f'Ignoring event for {source_path} because it does not exist.')
        return False
    if source_path.is_dir():
        logger.info(f'Ignoring event for {source_path} because it\'s a folder.')
        return False
    if not source_path.is_relative_to(app_config.source):
        logger.info(f'Ignoring event for {source_path} because it does not start with {app_config.source}.')
        return False
    if app_config.excludes.match(str(source_path)):
        logger.info(f'Ignoring {source_path} because it matches the exclusion pattern')
        return False
    return True

__all__ = [
    'link_source_to_destination',
    'prepare_logger',
    'source_path_is_qualified'
]

File: app/src/utilities.ts
import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';

import { AppConfig } from './AppConfig';
import { Logger } from './Logger';

export function initializeAppConfig(configPath: string = '/config/config.yaml'): AppConfig {
  const fileContents = fs.readFileSync(configPath, 'utf8'),
    config: any = yaml.load(fileContents);

  if (!config) {
    throw new Error('Configuration is empty.');
  }

  const apiKey = process.env.SYNCTHING_API_KEY || '';
  if (!apiKey) {
    throw new Error('No API key found.');
  }

  const host: string = process.env.SYNCTHING_HOST || '127.0.0.1',
    port: number = parseInt(process.env.SYNCTHING_PORT || '8384'),
    isHttps: boolean = ['1', 'true', 'yes'].includes((process.env.SYNCTHING_HTTPS || '0').toLowerCase()),
    sslCertFile: string | undefined = process.env.SYNCTHING_CERT_FILE;

  return new AppConfig(
    apiKey,
    host,
    port,
    parseInt(process.env.SYNCTHING_TIMEOUT || '60'),
    isHttps,
    sslCertFile,
    config.source || '/files/source/',
    config.destination || '/files/destination/',
    (config.filter || 'ItemFinished').split(','),
    new RegExp(config.excludes || '')
  );
}

function linkSourceToDestination(sourcePath: string, destinationPath: string, logger: Console): void {
  const destinationParent = path.dirname(destinationPath);
  if (!fs.existsSync(destinationParent)) {
    fs.mkdirSync(destinationParent, {recursive: true});
    logger.info(`Created parent directory ${destinationParent} for ${destinationPath}.`)
  }

  if (fs.existsSync(destinationPath)) {
    return;
  }

  try {
    fs.linkSync(sourcePath, destinationPath);
    logger.info(`Linked ${sourcePath} to ${destinationPath}`);
  } catch (error) {
    logger.error(`Error linking ${sourcePath} to ${destinationPath}:`, error);
  }
}

export function processSourcePath(sourcePath: string, appConfig: AppConfig, logger: Console): void {
  if (!sourcePathIsQualified(sourcePath, appConfig, logger)) {
    return;
  }

  const relative = path.relative(appConfig.source, sourcePath),
    destinationPath = path.join(appConfig.destination, relative);

  linkSourceToDestination(sourcePath, destinationPath, logger);
}

function sourcePathIsQualified(sourcePath: string, appConfig: AppConfig, logger: Console): boolean {
  if (!fs.existsSync(sourcePath)) {
    logger.info(`Ignoring event for ${sourcePath} because it does not exist.`);
    return false;
  }
  if (fs.lstatSync(sourcePath).isDirectory()) {
    logger.info(`Ignoring event for ${sourcePath} because it\'s a folder.`);
    return false;
  }
  if (!sourcePath.startsWith(appConfig.source)) {
    logger.info(`Ignoring event for ${sourcePath} because it does not start with ${appConfig.source}.`);
    return false;
  }
  if (appConfig.excludes && appConfig.excludes.test(sourcePath)) {
    logger.info(`Ignoring ${sourcePath} because it matches exclude pattern.`);
    return false;
  }
  return true;
}

function rotateOldLogs(logDirectory: string): void {
  try {
    if (!fs.existsSync(logDirectory)) {
      return;
    }

    const files = fs.readdirSync(logDirectory);
    const now = new Date();
    const tenDaysAgo = new Date(now.getTime() - 10 * 24 * 60 * 60 * 1000);

    files.forEach(file => {
      const match = file.match(/^linker-(\d{4})(\d{2})(\d{2})\.log$/);
      if (match) {
        const [, year, month, day] = match;
        const fileDate = new Date(parseInt(year), parseInt(month) - 1, parseInt(day));

        if (fileDate < tenDaysAgo) {
          const filePath = path.join(logDirectory, file);
          fs.unlinkSync(filePath);
          console.info(`Deleted old log file: ${file}`);
        }
      }
    });
  } catch (error) {
    console.error('Error rotating logs:', error);
  }
}

export function getLogger(channel?: string): Console {
  if (!process.env.WRITE_LOGS) {
    return console;
  }

  const now = new Date(),
    year = now.getFullYear(),
    month = String(now.getMonth() + 1).padStart(2, '0'),
    day = String(now.getDate()).padStart(2, '0'),
    logPathOverride = process.env.LOG_PATH?.trim(),
    logfilePath = (logPathOverride && logPathOverride.length > 0)
      ? logPathOverride
      : `/logs/linker-${year}${month}${day}.log`;

  const logDirectory = path.dirname(logfilePath);
  rotateOldLogs(logDirectory);

  return new Logger(logfilePath, channel);
}
